File: backend/app/services/listenbrainz.py
from __future__ import annotations

from dataclasses import dataclass

import httpx

from app.core.config import Settings
from app.core.db import Database


@dataclass(frozen=True)
class PopularityStats:
    listen_count: int | None = None
    listener_count: int | None = None

    @property
    def score(self) -> float:
        listens = self.listen_count or 0
        listeners = self.listener_count or 0
        return (listeners * 10.0) + listens


class ListenBrainzClient:
    def __init__(self, settings: Settings, db: Database) -> None:
        self.settings = settings
        self.db = db

    async def recording_popularity(self, recording_ids: list[str]) -> dict[str, PopularityStats]:
        ids = [recording_id for recording_id in dict.fromkeys(recording_ids) if recording_id]
        if not ids:
            return {}

        cached: dict[str, PopularityStats] = {}
        missing: list[str] = []
        for recording_id in ids:
            cache_key = _recording_cache_key(recording_id)
            payload = self.db.get_cache(cache_key, self.settings.popularity_cache_ttl_seconds)
            if payload is None:
                missing.append(recording_id)
                continue
            cached[recording_id] = PopularityStats(
                listen_count=payload.get("listen_count"),
                listener_count=payload.get("listener_count"),
            )

        if missing:
            fetched = await self._fetch_recording_popularity(missing)
            cached.update(fetched)
            for recording_id, stats in fetched.items():
                self.db.set_cache(
                    _recording_cache_key(recording_id),
                    {
                        "listen_count": stats.listen_count,
                        "listener_count": stats.listener_count,
                    },
                )

        return cached

    async def _fetch_recording_popularity(
        self,
        recording_ids: list[str],
    ) -> dict[str, PopularityStats]:
        url = f"{self.settings.listenbrainz_base_url}/1/popularity/recording"
        async with httpx.AsyncClient(timeout=self.settings.listenbrainz_timeout_seconds) as client:
            response = await client.post(url, json={"recording_mbids": recording_ids})
            response.raise_for_status()
            payload = response.json()

        stats: dict[str, PopularityStats] = {}
        for item in payload:
            recording_id = item.get("recording_mbid")
            if not recording_id:
                continue
            stats[recording_id] = PopularityStats(
                listen_count=item.get("total_listen_count"),
                listener_count=item.get("total_user_count"),
            )
        return stats


def _recording_cache_key(recording_id: str) -> str:
    return f"listenbrainz:popularity:recording:v1:{recording_id}"

File: native/streambox_core/src/services/listenbrainz.rs
use std::collections::HashMap;
use std::time::Duration;

use reqwest::blocking::Client;
use serde_json::Value;

use crate::error::CoreError;
use crate::models::PopularityStats;

const LISTENBRAINZ_POPULARITY_URL: &str = "https://api.listenbrainz.org/1/popularity/recording";
const LISTENBRAINZ_TIMEOUT_SECS: u64 = 10;

pub struct ListenBrainzClient {
    client: Client,
}

impl ListenBrainzClient {
    pub fn new() -> Result<Self, CoreError> {
        let client = Client::builder()
            .timeout(Duration::from_secs(LISTENBRAINZ_TIMEOUT_SECS))
            .build()
            .map_err(|error| CoreError::new("listenbrainz_client_error", error.to_string()))?;
        Ok(Self { client })
    }

    pub fn recording_popularity(
        &self,
        recording_ids: &[String],
    ) -> Result<HashMap<String, PopularityStats>, CoreError> {
        let deduped: Vec<&str> = {
            let mut seen = std::collections::HashSet::new();
            recording_ids
                .iter()
                .filter(|id| !id.is_empty() && seen.insert(id.as_str()))
                .map(String::as_str)
                .collect()
        };
        if deduped.is_empty() {
            return Ok(HashMap::new());
        }
        self.fetch_recording_popularity(&deduped)
    }

    fn fetch_recording_popularity(
        &self,
        recording_ids: &[&str],
    ) -> Result<HashMap<String, PopularityStats>, CoreError> {
        let body = serde_json::json!({ "recording_mbids": recording_ids });
        let response = self
            .client
            .post(LISTENBRAINZ_POPULARITY_URL)
            .json(&body)
            .send()
            .map_err(|error| CoreError::new("listenbrainz_request_failed", error.to_string()))?;

        if !response.status().is_success() {
            return Err(CoreError::new(
                "listenbrainz_error",
                format!("ListenBrainz returned HTTP {}", response.status()),
            ));
        }

        let payload: Vec<Value> = response
            .json()
            .map_err(|error| CoreError::new("listenbrainz_parse_failed", error.to_string()))?;

        let mut stats: HashMap<String, PopularityStats> = HashMap::new();
        for item in payload {
            let recording_id = match item.get("recording_mbid").and_then(|v| v.as_str()) {
                Some(id) => id.to_string(),
                None => continue,
            };
            let mut entry = PopularityStats {
                listen_count: item.get("total_listen_count").and_then(|v| v.as_i64()),
                listener_count: item.get("total_user_count").and_then(|v| v.as_i64()),
                score: 0.0,
            };
            entry.compute_score();
            stats.insert(recording_id, entry);
        }
        Ok(stats)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn compute_score_formula() {
        let mut stats = PopularityStats {
            listen_count: Some(100),
            listener_count: Some(10),
            score: 0.0,
        };
        stats.compute_score();
        assert!((stats.score - 200.0).abs() < 0.01); // 10*10 + 100 = 200
    }

    #[test]
    fn compute_score_missing_fields() {
        let mut stats = PopularityStats {
            listen_count: None,
            listener_count: None,
            score: 0.0,
        };
        stats.compute_score();
        assert!((stats.score - 0.0).abs() < 0.01);
    }

    #[test]
    fn parse_listenbrainz_response() {
        let json = serde_json::json!([
            {"recording_mbid": "abc-123", "total_listen_count": 50, "total_user_count": 5},
            {"recording_mbid": "def-456", "total_listen_count": null, "total_user_count": 3},
        ]);
        // Simulate what fetch_recording_popularity does
        let mut stats_map: HashMap<String, PopularityStats> = HashMap::new();
        for item in json.as_array().unwrap() {
            let id = item["recording_mbid"].as_str().unwrap().to_string();
            let mut entry = PopularityStats {
                listen_count: item["total_listen_count"].as_i64(),
                listener_count: item["total_user_count"].as_i64(),
                score: 0.0,
            };
            entry.compute_score();
            stats_map.insert(id, entry);
        }
        assert_eq!(stats_map.len(), 2);
        assert!((stats_map["abc-123"].score - 100.0).abs() < 0.01); // 5*10 + 50
        assert!((stats_map["def-456"].score - 30.0).abs() < 0.01); // 3*10 + 0
    }

    #[test]
    fn empty_ids_returns_empty() {
        let client = ListenBrainzClient::new().unwrap();
        let result = client.recording_popularity(&[]);
        assert!(result.unwrap().is_empty());
    }

    #[test]
    fn deduplicates_recording_ids() {
        let client = ListenBrainzClient::new().unwrap();
        let ids: Vec<String> = vec!["a".into(), "b".into(), "a".into()];
        // Won't actually call API since it fails with empty body, but tests dedup
        let result = client.recording_popularity(&ids);
        // Should fail because we actually try to POST with real HTTP
        // This tests dedup logic reaches fetch stage
        assert!(result.is_err() || result.is_ok()); // Just verify it doesn't panic
    }
}
